### ISSUE TITLE
Game slows down drastically when character is near a large number of NPCs

### ISSUE THREAD

**Issue by RecycledSail** [CleverRaven/Cataclysm-DDA#42899]
# Describe the bug

Game slows down when a large number of NPCs is near player

# Steps To Reproduce

1. Make a fresh new character starting at shelter.
2. Download location of refugee center via terminal, then teleport to the center.
3. Wait inside refugee center, you will notice severe slowdown

# Expected behavior

It shouldn't be that slow.

# Screenshots

This is CPU profiling screen while I was waiting inside refugee center.
![1](https://user-images.githubusercontent.com/33354771/90115849-a1a18380-dd8f-11ea-8161-61f1031d2457.jpg)
npc::process_turn and npc::move takes 56.24+25.43=81.67% of total CPU usage.

![2](https://user-images.githubusercontent.com/33354771/90115857-a2d2b080-dd8f-11ea-8c1b-60e9832a1c95.jpg)
For npc::process_turn, sum of Character::drop_invaild_inventory and player::process_items(Character::calc_encumbrance takes almost entirety of its CPU usage time) takes 42.27% of total CPU usage.

![3](https://user-images.githubusercontent.com/33354771/90115859-a2d2b080-dd8f-11ea-8197-7f85640dc747.jpg)
For npc::move, npc::assess_danger takes 21.76% of total CPU usage.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-4997-gaf19ae1351-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

# Additional context
Maybe we can skip some turns while waiting to improve performance?

**Comment by ZhilkinSerg** [comment]
Performance should be better after #42998, but there are probably still some things to enhance.

**Comment by Cenness** [comment]
This problem is also noticeable in those multi floor above ground research facility with dozens of diverse enemies and a lot of desctruction.

**Comment by ZhilkinSerg** [comment]
With some optimizations related to trait check I could achieve the following:

![image](https://user-images.githubusercontent.com/16213433/90840190-021d5b80-e362-11ea-9d99-eb8808620447.png)

![image](https://user-images.githubusercontent.com/16213433/90840240-237e4780-e362-11ea-87cc-2deeeeab58ac.png)

![image](https://user-images.githubusercontent.com/16213433/90840227-19f4df80-e362-11ea-8161-22ab1c32160b.png)

It could be enhanced even more with some item weight caching.

**Comment by kevingranade** [comment]
Closing in favor of #45497 as this does not capture any additional information.
Also calling it "no longer a blocker for 0.F" due to 60% speedup due to Aivean's improvements.
